onnxruntime/core/session/inference_session_utils.cc

```diff
@@ -0,0 +1,223 @@
+    LOGS(logger, ERROR) << "Unsupported execution_mode value in ORT config: " << value;
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported execution_mode value in ORT config: ", value);
+  }
+
+  LOGS(logger, INFO) << "Setting execution_mode to " << (value == 0 ? "Sequential mode" : "Parallel mode");
+  session_options.execution_mode = (value == 0 ? ExecutionMode::ORT_SEQUENTIAL : ExecutionMode::ORT_PARALLEL);
+  return Status::OK();
+}
+
+static Status SetGraphOptimizationLevel(SessionOptions& session_options,
+                                        int value,
+                                        const logging::Logger& logger) {
+  switch (value) {
+    case 0:
+      LOGS(logger, INFO) << "Setting graph_optimization_level to ORT_DISABLE_ALL";
+      session_options.graph_optimization_level = TransformerLevel::Default;
+      return Status::OK();
+
+    case 1:
+      LOGS(logger, INFO) << "Setting graph_optimization_level to ORT_ENABLE_BASIC";
+      session_options.graph_optimization_level = TransformerLevel::Level1;
+      return Status::OK();
+
+    case 2:
+      LOGS(logger, INFO) << "Setting graph_optimization_level to ORT_ENABLE_EXTENDED";
+      session_options.graph_optimization_level = TransformerLevel::Level2;
+      return Status::OK();
+
+    case 99:
+      LOGS(logger, INFO) << "Setting graph_optimization_level to ORT_ENABLE_ALL";
+      session_options.graph_optimization_level = TransformerLevel::MaxLevel;
+      return Status::OK();
+
+    default:
+      std::ostringstream message_stream;
+      message_stream << "Unsupported graph_optimization_level value in ORT config: " << value;
+
+      std::string message = message_stream.str();
+
+      LOGS(logger, ERROR) << message;
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, message);
+  }
+}
+
+static Status SetEnableProfiling(SessionOptions& session_options,
+                                 int value,
+                                 const logging::Logger& logger) {
+  if (value != 0 && value != 1) {
+    LOGS(logger, ERROR) << "Unsupported value for enable_profiling option: " << value;
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported value for enable_profiling option: ", value);
+  }
+
+  LOGS(logger, INFO) << "Setting enable_profiling to " << (value == 0 ? "false" : "true");
+  session_options.enable_profiling = (value == 0 ? false : true);
+  return Status::OK();
+}
+
+//---------------------------------------------------
+//--- end of session options related helpers ---
+//---------------------------------------------------
+
+//---------------------
+//--- end of local helpers ---
+//---------------------
+
+Status InferenceSessionUtils::ParseOrtConfigJsonInModelProto(const ONNX_NAMESPACE::ModelProto& model_proto) {
+  if (is_model_checked_for_ort_config_json_) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "The Model Proto has already been checked for the ORT config json.");
+  }
+
```

Comment:
nit: does it really matter if this is called twice? if not, whilst this is correct, it's extra lines of code that don't provide much value.

Reply:
My thoughts are the exact opposite - I think this check is pretty important and guards against an accidental call to parse the json again from potentially a different model. Since all this is done at session init time, I feel we can't go too wrong with having a few extra checks.

Reply:
Resolving this conversation as this is a nit. Will re-visit this later if you feel that it should be changed.